### ISSUE TITLE
[Bug]: LoRA `lora_expand` Triton kernel outputs NaN on Hopper (sm_90) with `block_n=128`, producing garbled LoRA output

### ISSUE THREAD

**Issue by n0whereRuoxi** [vllm-project/vllm#48590]
---

# === My current environment ===

<details>
<summary>Output of <code>python -m vllm.collect_env</code> (run inside the affected H20 container)</summary>

```text
==============================
        System Info
==============================
OS                           : Ubuntu 22.04.5 LTS (x86_64)
GCC version                  : (Ubuntu 11.4.0-1ubuntu1~22.04.3) 11.4.0
Libc version                 : glibc-2.35

==============================
       PyTorch Info
==============================
PyTorch version              : 2.11.0+cu130
CUDA used to build PyTorch   : 13.0

==============================
      Python Environment
==============================
Python version               : 3.12.13
Python platform              : Linux-4.18.0-372.32.1.90.po1.x86_64-x86_64-with-glibc2.35

==============================
       CUDA / GPU Info
==============================
Is CUDA available            : True
CUDA runtime version         : 13.0.88
GPU models and configuration : GPU 0: NVIDIA H20-3e   (Hopper, sm_90)
Nvidia driver version        : 580.105.08

==============================
Versions of relevant libraries
==============================
[pip3] flashinfer-python==0.6.12
[pip3] numpy==2.2.6
[pip3] torch==2.11.0+cu130
[pip3] torchvision==0.26.0+cu130
[pip3] transformers==5.12.0
[pip3] triton==3.6.0

vLLM Version                 : 0.23.0
```

Also reproduced on vLLM **0.24.0** and **0.25.0** — `vllm/lora/ops/triton_ops/lora_expand_op.py` and `kernel_utils.py` are byte-identical across 0.23/0.24/0.25, so upgrading does not fix it.

</details>

---

# === 🐛 Describe the bug ===

## Summary

On **NVIDIA Hopper (sm_90, e.g. H20/H100)**, serving a Qwen3-VL-8B-Instruct LoRA adapter with vLLM produces **completely garbled output** (e.g. repeated `!!!!!!`, or incoherent mixed-language/code-like tokens). The **base model without LoRA is perfectly fine**.

I traced this to the LoRA **`lora_expand`** Triton kernel producing **NaN**. `lora_shrink` is fine. The trigger is the default expand kernel config **`block_n = 128`** (single-slice path); forcing **`block_n = 32`** makes the kernel numerically correct and the garbling disappears.

It is **architecture-specific**: the exact same adapter runs correctly on **Blackwell (RTX 5070, sm_120)** and via **HF Transformers + PEFT** on the same GPU-family — only the vLLM Triton expand kernel on sm_90 is wrong.

## Transformers vs vLLM (per the issue template)

I ran the **same adapter** through plain **HF Transformers + PEFT** (`PeftModel.from_pretrained`) — output is **correct**. Only vLLM garbles. So this is a vLLM (Triton kernel) issue, not a modeling/transformers issue.

## What I ruled out (each tested, not guessed)

- vLLM version: 0.23.0 / 0.24.0 / 0.25.0 all reproduce (kernel code identical).
- `--enforce-eager` on/off: both reproduce.
- Attention backend: FlashAttn v2 / v3 / `TORCH_SDPA` all reproduce.
- Sampler backend: flashinfer sampler on/off, both reproduce.
- Quantization: bf16 / fp8 / bitsandbytes-4bit all reproduce.
- Virtualization: reproduced on **bare-metal H20** (no HAMi/vGPU).
- Stale `~/.triton/cache`: wiping it does **not** help; NaN reproduces on a fresh compile.

## Minimal, model-free reproduction

This calls only the LoRA Triton ops with synthetic tensors and compares against the torch reference (`vllm/lora/ops/torch_ops`):

```python
import torch
from vllm.lora.ops.triton_ops import lora_expand, LoRAKernelMeta
from vllm.lora.ops.torch_ops import bgmv_expand

dev = "cuda"
T, H, R, scaling = 1029, 4096, 64, 2.0   # num_tokens, hidden, rank, alpha/r
for dtype in (torch.bfloat16, torch.float16):
    x   = torch.randn(1, T, R, device=dev, dtype=dtype)          # shrink output [slices, tokens, rank]
    B   = (torch.randn(1, H, R, device=dev) * 0.02).to(dtype)    # lora_b [num_loras, hidden, rank]
    out = torch.zeros(T, H, device=dev, dtype=dtype)
    tlm = torch.zeros(T, dtype=torch.int32, device=dev)          # all tokens -> lora 0
    meta = LoRAKernelMeta.make(max_loras=1, max_num_tokens=T, device=dev)
    meta.prepare_tensors(tlm)

    lora_expand(x, [B], out, *meta.meta_args(T, False), offset_start=0, add_inputs=True)

    ref = torch.zeros(T, H, device=dev, dtype=dtype)
    bgmv_expand(x[0], B, ref, torch.zeros(T, dtype=torch.long, device=dev), add_inputs=True)

    print(dtype, "| lora_expand(triton) NaNs:", torch.isnan(out).sum().item(),
          "| bgmv_expand(torch) NaNs:", torch.isnan(ref).sum().item())
```

Observed on H20 (sm_90):

```text
torch.bfloat16 | lora_expand(triton) NaNs: 4480   | bgmv_expand(torch) NaNs: 0
torch.float16  | lora_expand(triton) NaNs: 36352  | bgmv_expand(torch) NaNs: 0
```

Only a subset of output elements are NaN (looks like a boundary N-tile), then it propagates through the forward pass → garbled decode.

## The trigger: `block_n = 128`

In `vllm/lora/ops/triton_ops/utils.py`, the default single-slice expand config is:

```python
"block_n": 64 if num_slices > 1 else 128,
```

Changing `128` → `32` yields **0 NaN** and matches the torch reference (max abs diff ~1e-3, normal bf16 rounding). So the fault is specific to the **`block_n=128` tile on sm_90** — pointing at a Triton codegen / memory-access bug at that tile size (possible out-of-bounds read on the boundary N-tile, or an MMA/shared-memory layout issue for a 128-wide tile on Hopper). Compare with #47303 (Triton MoE kernels producing NaN due to OOB reads) — similar flavor.

## Impact / fix verified

With either workaround, garbling disappears and downstream task accuracy is restored (0.63 base → ~0.90 LoRA, matching the Transformers+PEFT reference):

1. Set single-slice expand `block_n = 32` in `utils.py` (keeps native Triton, negligible perf hit, CUDA-graph safe). **Recommended.**
2. Fall back to a torch `lora_expand` on sm_90 (correct but slower; must be written CUDA-graph-safe or capture fails with `operation not permitted when stream is capturing`).

## Ask

- Is `block_n=128` for the expand kernel expected to be valid on sm_90? The NaN strongly suggests a tiling/masking/codegen bug at that config on Hopper.
- Could the default expand config avoid `block_n=128` on sm_90, or the boundary-tile masking be fixed?

Happy to provide full serve logs and the instrumentation details.

**Comment by elinacse** [comment]
@n0whereRuoxi Can I take it up ?

**Comment by Lei00764** [comment]
We reproduced and confirmed this bug with **Qwen3.6-35B-A3B MoE LoRA** on **H20 (sm_90)**, and the `block_n=32` fix resolves it.

## Reproduction

- **Model**: Qwen3.6-35B-A3B (`Qwen3_5MoeForConditionalGeneration`, hybrid GDN linear attention + full attention + sparse MoE)
- **LoRA adapter**: r=64, alpha=128, 780 keys, `target_parameters=[experts.gate_up_proj, experts.down_proj]`
- **GPU**: 8×H20-3e 144GB (Hopper sm_90), CUDA 13.0
- **vLLM**: 0.26.0 (cu130, torch 2.11.0+cu130)

## Symptoms

- **Short prompts**: LoRA works correctly (output differs from base model)
- **Long prompts** (>~128 tokens, e.g. 2335-char SciCode prompts): output becomes `!!!` (NaN), `/v1/completions` with `logprobs` returns `{"error":"Out of range float values are not JSON compliant: nan"}`

## Runtime Localization

Using `VLLM_ENABLE_V1_MULTIPROCESSING=0` + monkey-patching `Qwen3_5DecoderLayer.forward`:

1. **First NaN appears in layer 6** (GDN linear attention type), starting at **token 128**
2. All tokens from 128 onwards are NaN (346/474 tokens in a 474-token prefill)
3. `core_attn_out` (GDN attention core output) is NaN from token 128 — the NaN originates **inside** `torch.ops.vllm.qwen_gdn_attention_core`, not in the LoRA projection itself
4. LoRA weights are normal (max abs = 0.017, no NaN/inf) — **not a numerical overflow**
5. `linear_key_head_dim = 128`, `linear_value_head_dim = 128` — token 128 aligns with head_dim

The single-slice `lora_expand` kernel (used by GDN attention `out_proj` LoRA) gets `block_n = 128` from `get_lora_op_configs` (`64 if num_slices > 1 else 128`), which triggers the Hopper sm_90 NaN path.

## Fix

In `vllm/lora/ops/triton_ops/utils.py`, `get_lora_op_configs`, change the expand default:

```python
# Before:
"block_n": 64 if num_slices > 1 else 128,

# After:
"block_n": 64 if num_slices > 1 else 32,
```

## Verification

After the fix:
- Long prompts no longer produce NaN ✅
- 37-prompt LoRA effectiveness probe: 37/37 show output difference from base model ✅
- SciCode v51 training set (50 samples): **54.0% exact match, 95.1% avg similarity** — closely matching HF transformers + PEFT baseline (56.0% exact, 95.3% similarity)
- The ~2% gap is expected bf16 precision difference (vLLM Triton kernel vs torch matmul)

## Comparison with HF Transformers

Same adapter loaded via `PeftModel.from_pretrained` (transformers 5.5.3 + peft 0.19.1) produces **no NaN** on the same long prompts — confirming the bug is specific to vLLM's Triton `lora_expand` kernel, not the adapter weights.

## Note

We also found that `block_n=128` only triggers NaN for **single-slice** expand (GDN attention LoRA). Multi-slice expand (e.g. QKV projection) uses `block_n=64` and is unaffected. This is consistent with the issue title — the problem is specifically `block_n=128` on sm_90.
